content/blog/towards-continuous-performance-regression-testing.asciidoc

```
@@ -0,0 +1,594 @@
+We're going to discuss multiple examples for using JfrUnit to identify potential performance regressions.
+
+== Spotting Increased Memory Allocation
+
+At first, let's explore how to identify increased memory allocation rates.
+Typically, it's mostly library and middleware authors who are interested in this,
+since it can make a huge difference for a library such as Hibernate ORM whether a method invoked many times on a hot code path allocates a few bytes more or less memory.
+Allocating less memory means less work for the garbage collector,
+which in turn means those precious CPU cores of your machine can spend more cycles processing your actual business logic.
+
+But also for an actual application itself it can be beneficial to keep an eye on -- and systematically track -- object allocations,
+as regressions there lead to increased GC pressure,
+and in turn eventually to higher latencies and reduced throughput.
+
+The key for tracking object allocations with JFR are the `jdk.ObjectAllocationInNewTLAB` and `jdk.ObjectAllocationOutsideTLAB` events,
+which are emitted when
+
+* an object allocation triggered the creation of a new thread-local allocation buffer (TLAB) 
+* an object got allocated outside of the threads TLAB
+
+[NOTE]
+.Thread-local allocation buffers (TLAB)
+====
+When creating new object instances on the heap,
+this primarily happens via _thread-local allocation buffers_.
+A TLAB is a pre-allocated memory block that's exclusively used by a single thread.
+Creating new objects within a TLAB can happen without costly synchronization with other threads.
+Once a thread's current TLAB capacity is about to be exceeded by a new object allocation,
+a new TLAB will be allocated for that thread.
```

hum I don't remember the limit defaults, but I think it's a bit misleading to let people think it will just keep growing, as the whole point of our most important optimisations is to not exceed the limit.

---

I'm not quite following on that one; what is the "it" in "it will just keep growing"? The TLAB doesn't grow, it will be used up by allocations, and when the next allocation doesn't fit into it, the thread will get a new TLAB for the next set of allocations. You won't avoid that, unless you do zero allocations or allocate in your own self-managed `byte[]`.